### PR TITLE
Key API updates and bug fixes.

### PR DESCRIPTION
**Pull request by btimby**
This PR makes a few small changes to the API client.
- The trailing slash after key name bothered me so I removed it. The slash after username is still necessary.
- PUT method is now used to create AND update keys. This means we save a round-trip to check if something exists before saving it.
- I added caching for keys, this benefits downstream in several ways, I am able to remove some caching. Round trips are reduced.
- Deep copy global options when making an HTTP request so that they are not polluted.

### CODE REVIEW

**Comment by btimby** [comment]
Closes https://github.com/smartfile/client-node/issues/53

**Comment by btimby** [comment]
Closes https://github.com/smartfile/client-node/issues/52

File: lib/keys.js [hunk 1]
```diff
@@ -33,63 +51,103 @@ class Keys {
...
+
+  find(givenKey, callback) {
+    this.list((e, keys) => {
+      if (e) {
+        if (callback) callback(e);
         return;
       }
 
-      callback(e, json);
+      for (let i = 0; i < keys.length; i++) {
```
Comment (cabarnes) [inline lib/keys.js:138]: Couldn't this be a `forEach`?

Comment (btimby) [reply lib/keys.js:138]: Not the same, I am returning once I find the key. In `forEach()` return just advances to the next item.

**Pull request merged**